File: src/main/java/org/example/bankramenserver/domain/auth/service/JwtService.java
package org.example.bankramenserver.domain.auth.service;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.example.bankramenserver.domain.auth.exception.ExpiredTokenException;
import org.example.bankramenserver.domain.auth.exception.InvalidTokenException;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.util.Date;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

@Service
public class JwtService {

    private final SecretKey secretKey;
    private final long accessExpSeconds;
    private final long refreshExpSeconds;

    public JwtService(
            @Value("${jwt.secretKey}") String secret,
            @Value("${jwt.accessExp}") long accessExpSeconds,
            @Value("${jwt.refreshExp}") long refreshExpSeconds
    ) {
        this.secretKey = Keys.hmacShaKeyFor(secret.getBytes());
        this.accessExpSeconds = accessExpSeconds;
        this.refreshExpSeconds = refreshExpSeconds;
    }

    public String generateAccessToken(UUID userId) {
        return Jwts.builder()
                .subject(userId.toString())
                .claim("type", "access")
                .id(UUID.randomUUID().toString())
                .issuedAt(new Date())
                .expiration(new Date(
                        System.currentTimeMillis()
                                + TimeUnit.SECONDS.toMillis(accessExpSeconds)
                ))
                .signWith(secretKey)
                .compact();
    }

    public String generateRefreshToken(UUID userId) {
        return Jwts.builder()
                .subject(userId.toString())
                .claim("type", "refresh")
                .id(UUID.randomUUID().toString())
                .issuedAt(new Date())
                .expiration(new Date(
                        System.currentTimeMillis()
                                + TimeUnit.SECONDS.toMillis(refreshExpSeconds)
                ))
                .signWith(secretKey)
                .compact();
    }

    public UUID validateAccessToken(String token) {
        Claims claims = parse(token);

        if (!"access".equals(claims.get("type"))) {
            throw InvalidTokenException.EXCEPTION;
        }

        return UUID.fromString(claims.getSubject());
    }

    public UUID validateRefreshToken(String token) {
        Claims claims = parse(token);

        if (!"refresh".equals(claims.get("type"))) {
            throw InvalidTokenException.EXCEPTION;
        }

        return UUID.fromString(claims.getSubject());
    }

    private Claims parse(String token) {
        try {
            return Jwts.parser()
                    .verifyWith(secretKey)
                    .build()
                    .parseSignedClaims(token)
                    .getPayload();

        } catch (ExpiredJwtException e) {
            throw ExpiredTokenException.EXCEPTION;
        } catch (Exception e) {
            throw InvalidTokenException.EXCEPTION;
        }
    }
}

File: src/main/kotlin/org/example/balogserver/domain/auth/service/JwtService.kt
package org.example.balogserver.domain.auth.service
import io.jsonwebtoken.Claims
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.example.balogserver.domain.auth.exception.ExpiredTokenException
import org.example.balogserver.domain.auth.exception.InvalidTokenException
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.Date
import java.util.UUID
import java.util.concurrent.TimeUnit
import javax.crypto.SecretKey
@Service
class JwtService(
    @param:Value("\${jwt.secretKey}") secret: String,
    @param:Value("\${jwt.accessExp}") private val accessExpSeconds: Long,
    @param:Value("\${jwt.refreshExp}") private val refreshExpSeconds: Long,
) {
    private val secretKey: SecretKey = Keys.hmacShaKeyFor(secret.toByteArray())

    fun generateAccessToken(userId: UUID) = generateToken(userId, "access", accessExpSeconds)
    fun generateRefreshToken(userId: UUID) = generateToken(userId, "refresh", refreshExpSeconds)
    fun validateAccessToken(token: String) = validate(token, "access")
    fun validateRefreshToken(token: String) = validate(token, "refresh")

    private fun generateToken(userId: UUID, type: String, expires: Long) = Jwts.builder()
        .subject(userId.toString()).claim("type", type).id(UUID.randomUUID().toString())
        .issuedAt(Date()).expiration(Date(System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(expires)))
        .signWith(secretKey).compact()

    private fun validate(token: String, type: String): UUID {
        val claims = parse(token)
        if (claims["type"] != type || claims.expiration == null) throw InvalidTokenException.EXCEPTION
        val subject = claims.subject ?: throw InvalidTokenException.EXCEPTION
        return try {
            UUID.fromString(subject)
        } catch (_: IllegalArgumentException) {
            throw InvalidTokenException.EXCEPTION
        }
    }

    private fun parse(token: String): Claims = try {
        Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token).payload
    } catch (_: ExpiredJwtException) {
        throw ExpiredTokenException.EXCEPTION
    } catch (_: Exception) {
        throw InvalidTokenException.EXCEPTION
    }
}
